### ISSUE TITLE
Collateral API extension

### ISSUE THREAD

**Issue by maektwain** [upscaletech/finocial#8]
## Knowledge

For Collateral management you need to understand the logic like this 

The current functionality for loan collaterals is minimal and tightly linked to loan creation (you are essentially storing details of loan collaterals along with their value while creating a loan)

For organizations which work extensively on collateral based loans (say gold loans etc) , this wouldn’t work well given some of the common business scenarios

For ex:  A walk-in customer would want to know the details of the loans he could get based on his collateral (so you are not really looking to create this customer or his loan account in your MIS yet)

At  high level, we would probably need the following functionality
1. A set of API’s which define the collateral definitions, base price and quality standards (if applicable)

So for example, If I were to create a Collateral definition for gold, I would set the base quantity as 1 gram and its price as say 30,000 Rupees
I would then go ahead and define the different quality standards for this Collateral, i.e 24K gold which is with the base price of the collateral, 22k gold which is worth 85% of the collateral price and so on.
As the market rate of gold changes, I would update the base price and the same would be reflected for all quality standards which are derived based on the base price
1. Loan product definition should allow capturing permissable Collateral types and the max asset percentage that can be sanctioned as the loan amount
2. We would also need functionality which would allow pledging a combination of different Collaterals and ascertain their market value.
   Based on the total value of all collaterals in a single pledge, you would then determine the available loans and their max permissible disbursal amounts. Using this, a customer decided if he is interested in taking any of the loans...

The details you would store for each collateral in a pledge would be similar to:
Gold with quality standard 22k and a gross weight of 24 grams and a net weight of 20 grams (calculated by removing any stone weight etc) would have a system calculated value, based on this a user can provide an alternate value which is lesser than the system value.
The user defined value overrides the system calculated value and is used for all calculations...

4) After creating a customer, you would need API’s for associating a pledge (where a pledge contains details of the collaterals he is pledging for a loan) with a customer and the ability to then create a loan account linked to this pledge

Since the sequence of activities in purely collateral based organizations (where most customers are walk-in customers) is usually different from those of a regular organization (i.e you capture details of the pledge first and then if the customer agrees to you loan terms, you then proceed to create customer accounts, loan account etc)
It might make sense to allow saving pledges first (independent of the customer) and then creating customers associated with these pledges etc to better suit the real world workflows…...
1. The main packages to look out for are
   org.apache.fineract.portfolio.collateral and
   org.apache.fineract.portfolio.loanaccount .
2. Basically we want that the functions such as createcollateral,
   retrievecollateral, updatecollateral etc. be independent of loan
   specific details such as loanID etc.
3. We need new set of API's which make use of the collateral using
   their unique collateralID and attach them to the individual loan
   objects. This way both the loan and collateral become loosely coupled.
4. We need collateral value calculation engines. This engine should be a robust engine making use of
   the current market prices for assets such as gold, securities etc.

@mit2u can you provide us more information about the Collateral Functioning which is going to get developed for Upscale Finocial 

`POST /collateral`

``` json

{}
```

**Comment by mit2u** [comment]
We can have two modules running ; one used for defining a collateral product and other for storing the information of collateral pledging.

For Example : say a bank is setting up a two wheeler loan, he will set the loan product via a normal procedure.And in the collateral definition module similar to loan product definition, bank can define that this loan product is for such and such collateral.Collateral definition should include settings like
-Valuation
-Depreciation
-product with corresponding definitions
 Code Values can be [ Bike : manufacturer, Model, ExshowroomPrice, City, State, Make, TemporaryNo ; 
 Car: manufacturer,Model,ExshowroomPrice,City,state,Make,TemporaryNo ]
-LTV
-Insurance : corresponding definitions
  Code Values can be [First Party : Insurer-----incomplete----

**Comment by maektwain** [comment]
Yeah that is the initial plan because this will allow us to configure the collateral for every loan based on the loan Product or maybe we can define the configurability of collateral inside the Loan Product Api @anuragmath  what do you think of it

@jdc91  Provide your input on this

**Comment by jdc91** [comment]
Hi @maektwain  and @mit2u! Sorry Ive been in the dark for a few days. So on this requirement for collateral.
As I can understand we will be 

- Denormalizing the existing loan schema
- Refactor the CRUD for existing loans module.
- Create a new decoupled service for collateral altogether.

So the main task will be this comparison of Loans and Collaterals to derive valuations of both and see which of the loan products are eligible for one or more pledges?

Am I right?

Have a few clarifications though.
- Are we allowing highly customizable collaterals? For example can I offer up my mac book as the collateral? Is the real world operation going to reflect this? Will we allow collaterals to be created every-time a new customer comes up with some valuable product to pledge [if that collateral does not exist of course].
Seems like this collateral module is going to be pretty huge. Its good if we can maybe start off with a schema design and changes to existing schema.

Also @mit2u didnt get your 

> product with corresponding definitions
> Code Values can be [ Bike : manufacturer, Model, ExshowroomPrice, City, State, Make, TemporaryNo ; 

Part. 
What do you mean by this corresponding definition?

**Comment by jdc91** [comment]
Brace yourselves Team looks like we are disrupting a massive market with this single module ;)
> The global collateral management market is worth in excess of €10 trillion. 
> http://www.clearstream.com/blob/10620/e5bf3b589c8f3ff6afd19166f9d53d3b/accenture-collateral-report-pdf-data.pdf

**Comment by theupscale** [comment]
@jdc91 

About clarifications 

1. Yes have to these operations should highlight real world scenarios 
2. Yes we can first start with the model of module configuration schema

**Comment by mit2u** [comment]
@jdc91  yes collateral module will be huge, 
Yes their will be customization per say for a laptop collateral product properties taken are Processor*
RAM*
Hard Disk*
Make and type
Model
Dealer
Price

from my side some things are still pending ...that what more fields and properties can be taken.If you got the idea you can start creating the resource else wait , I will share the document on drive in 1-2 days.

**Comment by jdc91** [comment]
Hi Team,
Sorry for the delay, but came up with this ER for the DB Schema
Please provide feedback!
![image](https://cloud.githubusercontent.com/assets/3422709/21669928/fc1cd30e-d335-11e6-89da-bbc8675a555f.png)

**Comment by jdc91** [comment]
@mit2u I guess we have to run more revisions. Can you please resend that document you made on this module?

**Comment by mit2u** [comment]
@jdc91 Check your mail, I have sent a use case for managing collateral. Schema you made needs some more customization. Gold is only one use case. Add a collateral_type_ID int which links to multiple types of collateral.

**Comment by jdc91** [comment]
From that Doc,
Arent all these **Collateral properties** rather than pledge properties?.... A pledge can contain multiple collateral's right? Like I can pledge my jewelry and laptop ?

> Pledging Properties:
> Gold Loan
> Gold weight
> Type of jewellery
> No of units
> Gold value = market price
> Price
> Purity %
> Two/Four Wheeler Loan
> Manufacturer
> Model
> Version : default standard
> Make
> Colour
> RTO
> Dealer
> On road price and its breakup
> TVs
> Model
> Manufacturer
> Type : LCD/LED
> Screen Size
> Dealer
> Price
> Laptop
> Processor*
> RAM*
> Hard Disk*
> Make and type
> Model
> Dealer
> Price
> SME
> Property description
> Location 
> Valuation amount
> Name of valuer
> Title documents
> Units 
>

**Comment by mit2u** [comment]
As you find fit. 
But where should we define a type id?  For Mac = 1 , For gold =2.....
So that later we can sort all products by their types.

**Comment by jdc91** [comment]
Yeah this is a complex relationship... almost as if this bit of the db is **unstructured**! 
Btw, In the design collateral Id is <==> the type id you are talking about.
These collateral attributes will come in a different table.
![image](https://cloud.githubusercontent.com/assets/3422709/21748658/f2a23a40-d5af-11e6-87c8-35e140e1fd39.png)

WDYT?
